### PR TITLE
Refactor Cinema Hall into MonteWay

### PR DESCRIPTION
**Pull request by Minpeople**
**Cinema Hall --- MONTE WAY**
1. I created representers (single and multiple, but I thought earlier about creating just method "multiply").
2. I created use cases (create, delete, update and I moved there my "`generate seats`" file).
3. I created repository (when I was creating repositories for other models I've started thinking about DRY, because many repositories have the same methods so maybe create one general repository and just inherit from it and add extra methods if needed)
4. I updated `seeds` and `cinema_halls_controller`

### CODE REVIEW

File: app/controllers/cinema_halls_controller.rb [hunk 1]
```diff
@@ -4,12 +4,12 @@ class CinemaHallsController < ApplicationController
   include JSONAPI::Fetching
 
   def index
-    @cinema_halls = CinemaHall.all
+    @cinema_halls = CinemaHalls::Representers::Multiple.new.call
```
Comment (baniak66) [inline app/controllers/cinema_halls_controller.rb:7]: You should pass to the initializer collection of cinema_halls. For example:
```
cinema_halls = CinemaHall.all
CinemaHalls::Representers::Multiple.new(cinema_halls).call
```
Please check do you still need jsonapi if you have a representer.

File: app/controllers/cinema_halls_controller.rb [hunk 2]
```diff
@@ -36,7 +37,7 @@ def update
   end
 
   def destroy
-    @cinema_hall = CinemaHall.find(params[:id])
+    @cinema_hall = CinemaHalls::UseCases::Delete.new(id: params[:id]).call
```
Comment (baniak66) [inline app/controllers/cinema_halls_controller.rb:40]: You don't need to assign the result of use_case to `@cinema_hall`

File: app/controllers/cinema_halls_controller.rb [hunk 3]
```diff
@@ -36,7 +37,7 @@ def update
...
     @cinema_hall.destroy
```
Comment (baniak66) [inline app/controllers/cinema_halls_controller.rb:41]: The previous line should delete cinema_hall as I understand.

File: app/controllers/cinema_halls_controller.rb [hunk 4]
```diff
@@ -21,13 +21,14 @@ def create
   end
 
   def show
-    @cinema_hall = CinemaHall.find(params[:id])
+    @cinema_hall = CinemaHalls::Representers::Single.new(id: params[:id]).call
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb:24]: You should pass here a specific object, not an ID. The responsibility of `Representer` is to `represent` some object which is already pushed to this class.

File: app/controllers/cinema_halls_controller.rb (hunk 2 above)
Comment (Minpeople) [reply app/controllers/cinema_halls_controller.rb:40]: @baniak66, I just want to confirm before commit
You mean that I can change this line into:  `@cinema_hall = CinemaHalls::UseCases::Delete.new.call`  ?

**Comment by wmaciejak** [comment]
@Minpeople - general thing which was also noticed by @baniak66 

>Please check do you still need jsonapi if you have a representer.

If you already using JSON:API - you have serializers in your application. Responsibility of serializers is to represent objects in some standardized way of doing it. Responsibility of representers is exactly the same, so you have to:
1. Remove JSON:API at all
or
2. Not use representers at all

I do not have any preferences which one way you should take, but IMO if you have already standardized API, you don't have to create new Representers all the time. In other words - skip representers in your implementation.

File: app/controllers/cinema_halls_controller.rb (hunk 3 above)
Comment (wmaciejak) [reply app/controllers/cinema_halls_controller.rb:41]: Exactly. Deletion logic should be moved to UseCase class.

**Comment by wmaciejak** [comment]
> @Minpeople - general thing which was also noticed by @baniak66
> 
> > Please check do you still need jsonapi if you have a representer.
> 
> If you already using JSON:API - you have serializers in your application. Responsibility of serializers is to represent objects in some standardized way of doing it. Responsibility of representers is exactly the same, so you have to:
> 
> 1. Remove JSON:API at all
>    or
> 2. Not use representers at all
> 
> I do not have any preferences which one way you should take, but IMO if you have already standardized API, you don't have to create new Representers all the time. In other words - skip representers in your implementation.

Except for this, you have to rework a bit controllers. Anyway, you did a huge bunch of good work with responsibility separation. Good job 👌

File: app/controllers/cinema_halls_controller.rb (hunk 2 above)
Comment (wmaciejak) [reply app/controllers/cinema_halls_controller.rb:40]: When you deleting objects, you will probably don't need result of this action, so you don't need to assign result of this class to `@cinema_hall`. We meant that you should change this line into just:
`CinemaHalls::UseCases::Delete.new(id: params[:id]).call`

File: app/controllers/cinema_halls_controller.rb [hunk 5]
```diff
@@ -4,40 +4,39 @@ class CinemaHallsController < ApplicationController
...
   rescue ActiveRecord::RecordNotFound => e
-    render jsonapi: { error: e.message }, status: :not_found
+    render json: { error: e.message }, status: :not_found
   end
 
   def update
-    if @cinema_hall.update(cinema_hall_params)
-      render jsonapi: :show, status: :ok
+    cinema_hall = CinemaHalls::UseCases::Update.new(id: params[:id], params: cinema_hall_params).call
+    if cinema_hall.update(cinema_hall_params)
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb:31]: You don't need to update object twice. You already updated cinema_hall in 30 line

Comment (Minpeople) [reply app/controllers/cinema_halls_controller.rb]: @wmaciejak Updated 😅

File: app/controllers/cinema_halls_controller.rb [hunk 6]
```diff
@@ -4,40 +4,38 @@ class CinemaHallsController < ApplicationController
...
-    render jsonapi: @cinema_halls
+    cinema_halls = CinemaHall.all
+    render json: CinemaHalls::Representers::Multiple.new(cinema_halls).call
   end
 
   def create
-    @cinema_hall = CinemaHall.new(cinema_hall_params)
+    cinema_hall = CinemaHalls::UseCases::Create.new(params: cinema_hall_params).call
 
     if @cinema_hall.save
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb:14]: Here is the same situation. You trying to save object itself twice. You can try to rewrite it to:
```ruby
    cinema_hall = CinemaHalls::UseCases::Create.new(params: cinema_hall_params).call
    
    if cinema_hall.valid?
      render...
    else
      render ...
    end
```

Or even better:

```ruby
    cinema_hall = CinemaHalls::UseCases::Create.new(params: cinema_hall_params).call
    
    render success
  rescue
    render failure
  end
```
But in second approach you will have to properly work with errors

File: app/controllers/cinema_halls_controller.rb [hunk 7]
```diff
@@ -4,40 +4,40 @@ class CinemaHallsController < ApplicationController
...
-      render jsonapi: @cinema_halls.errors,
+      render json: cinema_halls.errors,
              status: :unprocessable_entity
     end
   end
 
   def show
-    @cinema_hall = CinemaHall.find(params[:id])
-    render jsonapi: @cinema_hall
+    cinema_hall = CinemaHalls::Representers::Single.new(cinema_hall).call
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb:24]: I'm pretty sure that you don't have access to `cinema_hall` variable which have to passed to representer.

File: app/controllers/cinema_halls_controller.rb [hunk 8]
```diff
@@ -4,40 +4,40 @@ class CinemaHallsController < ApplicationController
   include JSONAPI::Fetching
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb:4]: If you removed JSON:API, please remove usage of this module also

File: app/controllers/cinema_halls_controller.rb [hunk 9]
```diff
@@ -21,9 +20,8 @@ def create
   end
 
   def show
-    cinema_hall = CinemaHalls::Representers::Single.new(cinema_hall).call
-  rescue ActiveRecord::RecordNotFound => e
-    render json: { error: e.message }, status: :not_found
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb]: Rescue here wasn't a bad idea :) just fyi

**Review by wmaciejak — APPROVED** [review]

**Pull request merged**